src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -119,6 +99,142 @@ type accumulateWorker struct {
 	numErrors      int
 }
 
+type seriesBlock struct {
+	resolver bootstrap.SeriesRefResolver
+	block    block.DatabaseBlock
+}
+
+type readSeriesBlocksWorker struct {
+	dataCh      chan seriesBlock
+	reader      fs.DataFileSetReader
+	shard       uint32
+	accumulator bootstrap.NamespaceDataAccumulator
+	blocksPool  block.DatabaseBlockPool
+	blockStart  time.Time
+	blockSize   time.Duration
+	nsCtx       namespace.Context
+}
+
+func (w *readSeriesBlocksWorker) readSeriesBlocks(ctx context.Context) error {
+	defer close(w.dataCh)
+	for {
+		id, tags, data, expectedChecksum, err := w.reader.Read()
+		if err != nil && !errors.Is(err, io.EOF) {
+			return err
+		}
+		if errors.Is(err, io.EOF) {
+			break
+		}
+
+		dbBlock := w.blocksPool.Get()
+		dbBlock.Reset(w.blockStart, w.blockSize,
+			ts.NewSegment(data, nil, 0, ts.FinalizeHead), w.nsCtx)
+
+		// Resetting the block will trigger a checksum calculation, so use
+		// that instead of calculating it twice.
+		checksum, err := dbBlock.Checksum()
+		if err != nil {
+			return err
+		}
+		if checksum != expectedChecksum {
+			return fmt.Errorf("checksum for series: %s was %d but expected %d",
+				id, checksum, expectedChecksum)
+		}
+
+		res, owned, err := w.accumulator.CheckoutSeriesWithoutLock(w.shard, id, tags)
+		if err != nil {
+			if !owned {
+				// Skip bootstrapping this series if we don't own it.
+				continue
+			}
+			return err
+		}
+
+		w.dataCh <- seriesBlock{
+			resolver: res.Resolver,
+			block:    dbBlock,
+		}
+
+		id.Finalize()
+		tags.Close()
+
+		select {
+		case <-ctx.Done():
```

Comment:
This might be expensive to check on every single series, perhaps we can check this every N series? Say every 1024?

Reply:
done: https://github.com/m3db/m3/pull/3316/files#diff-8ac680b720a386924c879c03fad30c39581359a9b8fcad59814ccab0399703b0R164